app/models/report.rb

```diff
@@ -74,11 +74,15 @@ def <=>(other)
   # Expire reports based on time and status
   # Defaults to expire reports older than a week regardless of the status
   # This method will IS very slow, use only from rake task.
-  def self.expire(conditions, batch_size, sleep_time)
+  def self.expire(conditions, batch_size, sleep_time, keep_records)
     timerange = conditions[:timerange] || 1.week
     status = conditions[:status]
     created = (Time.now.utc - timerange).to_formatted_s(:db)
     logger.info "Starting #{to_s.underscore.humanize.pluralize} expiration before #{created} status #{status || 'not set'} batch size #{batch_size} sleep #{sleep_time}"
+    if keep_records > 0 && ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('postgresql')
+      logger.debug "Deleting logs up to last #{keep_records} records to speed up expiration process"
+      ActiveRecord::Base.connection.execute("DELETE FROM logs WHERE logs.id + #{keep_records} < (SELECT last_value FROM logs_id_seq)")
```

Comment:
I may be missing something, but I think this might delete logs for reports that are kept in cases when keep_record is low and the latest reports have a (relatively) high number of logs. Consider the following:

There are 10 reports, each has 10 logs. Reports have ids 1..10, logs have ids 1..100. 
`Report.expire({}, 1000, 0, 5)` will generate `DELETE FROM logs WHERE logs.id + 5 < 100`, so logs with ids 1..94 are deleted, but what if the last report is kept?

Reply:
The idea behind this is to quickly delete from the biggest logs table without join. We can't use any data from associated records, so the only thing the query can use is database ID. This was designed deliberately in a way that after deletion there can be report(s) left with either no content (zero logs = lines) or even "broken" report with just a portion of logs (lines). I believe this is the best I can achieve to boost times of expiration process.

Reply:
The only issue I can think of is when there are gaps in ID sequence. For example when somebody expired a bunch of reports so it created gaps. However since there is no option to delete reports in the UI/API, this should not happen. It could actually happen tho when expiration task is not catching up (it is slow) so there is one huge gap in IDs - only the initial run after ugprade could actually delete more records than it should. But then it would stabilize as this patch aims to speed up expiration process to catch up with new records.

Reply:
I think you can delete reports from [UI](https://github.com/theforeman/foreman/blob/develop/app/controllers/config_reports_controller.rb#L34) or [API](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/config_reports_controller.rb#L50) to create gaps.

It would not occur to me as a user that a task to expire old reports can corrupt the reports I want to keep, so I would not check that some logs in my report might be missing.

If this is an intentional behaviour, I would suggest we have it only as opt-in with a huge warning.

Reply:
Based on what @legooolas [suggested](https://github.com/theforeman/foreman/pull/7072#issuecomment-539977711), delete logs for each batch based on max report id

```ruby
query = where(cond).reorder('').limit(batch_size)
max_id = query.maximum('id')
ActiveRecord::Base.connection.execute("DELETE FROM logs WHERE logs.report_id <= #{max_id}")
```

I haven't tested, it probably will not be as fast due to calling delete for each batch though...

Reply:
@lzap was this incorporate in some way? did you test if deletion based on logs.report_id would be performing much worse?

Reply:
I would be ok to have this behaviour as it corrupts only reports on the very edge and fastens the proces. I believe the proposed solution will be slower as the where condition will slow it down.
There is a point though, that I may want to keep old records of some type even really old ones - forever maybe - and this will delete them. Note, that proposed solution would not solve this though. It would just keep the data integrity.

Reply:
What about:
```suggestion
      max_report_id = ActiveRecord::Base.connection.select_value("SELECT MAX(logs.report_id) as max_report_id FROM logs WHERE logs.id + #{keep_records} < (SELECT last_value FROM logs_id_seq)")
      ActiveRecord::Base.connection.execute("DELETE FROM logs WHERE report_id < #{max_report_id}")
```
It would keep the consistency and keep slightly less log records than the `keep_records`. The problem with the conditial preservation of records is not solved though.

Reply:
Well in this scenario there is a problem with hosts which no longer checks in - customers often have these (e.g. host is shutdown due to maintainance). Its reports gets older and older as new reports (and logs) come in. This will eventually create huge "gap" between the last report (shutdown host) and incoming (fresh) reports to a degree of months or even a year. This will render the PR useless - nothing will be ever deleted.

I've changed my mind and I made this opt-in with a big warning in the rake task help. We will write a KBASE article and only recommend this to customers who suffers from huge logs tables. They will be warned, I don't think we need to do any heuristics in this case.

Reply:
Wait let me think about it, I think it might be a possibility.

Reply:
Yeah, I understand and I am ok if you don't want my solution, but it would at least keep the integrity (eighter all logs, or no logs are kept for the report) and shouldn't be much slower than your solution (we have index on `report_id` so the `MAX` should be fast.